Improve creature leashing behaviour.

zana244 opened the pull request:
New year's gift! :D
Using a few patches from AC, which are themselves based on the creature leashing logic from vmangos.

Features:
Creatures which get called for assistance on the initial aggro pull will share the same leash timer with the initial target.
Link the leash timers of creatures and their pets, so they will keep each other in combat when either updates the leash timer.
Implement a rudimentary evade for the pets of creatures when the owner creature evades.
If a mob is not moving for about 5 seconds, while being able to attack the player, it will reupdate his leash timer.
Leash timer remains the same -> 12 seconds.

Extra -- Properly fix spells with SPELL_ATTR1_DISMISS_PET which I had broken in https://github.com/Project-Epoch/TrinityCore/pull/88
My initial fix works for players, the initial TC state worked only for creatures. Incorporate both to handle players/creatures.

Shortcomings:
1. If the players tryhard, they can still separate linked creatures; even if they share the same leash timer,
This is due to the check in CanCreatureAttack()
            return victim->IsInDist(&m_homePosition, dist);

Even if the leash timer expired, the creatures cannot reset if the victim (the player generally) is still within a HUGE radius around their home position.  --> if you skirt outside the radius of a creature's home position, but within the radius of a linked creature, the former will reset while the latter will still attack you.

vmangos solves this (I think ? maybe not this particular issue) by having an m_leashDistance assigned to each creature (most likely a config value) and checking it in Creature::Update --> presumably to allow some creatures to leash further than others.

2. When the owner creature evades, the pet is not in a true UNIT_STATE_EVADE --> you can attack it back and it will reaggro;
generally not a big issue, since most creatures recall their pet and dismiss this one when they reach their home position. Did not mess with UNIT_STATE_EVADE (yet), since there might be creatures who do not call their pet (or better yet, dismiss it), such that there is noone in the core to remove the UNIT_STATE_EVADE flag on the pet, rendering it permanently in an EVADE state.
This behaviour was present even before this PR, but at least now both the pet and the owner evade at the same time.

3. A bit unsure about the fanning aspect of mobs, and if they should reset the leash timer or not during this movement. I reverted the reset timer due to movement to around 5 seconds (vmangos value), since the one in AC (1.5 sec) would lead to a lot of false positives where the leash timer would reupdate while the mob would catch up to the player and attack him for < 1 sec.
Unsure if it's related to the i_recalculateTravel (in AC) and m_bRecalculateTravel (in vmangos) variables in the MovementGenerators (I commented out the code in this PR).

To differentiate between the initial CallAssistance() issued on initial pull and the CallAssistance() while mobs are chasing you, implement another bool variable which is true only for the first CallAssistance --> it allows us to know whether we should link the leash timers or not.

T1ti commented:
For reference if you want to implement it : 

> Mobs chase you for different lengths of time depending on level.
> 1-29: ~11 sec
> 30-39: ~12 sec
> 40-44: ~13 sec
> 45-49: ~14 sec
> 50-60: ~15 sec

zana244 commented:
I'll take your word for it, implemented a simple if/else case for it; not sure if the uint32 return value has anything to do with time_t addition

Reverted the linked leashes for creatures and pets, it seems that even though they had separate leashes, they both update at more or less the same time; less code added by me is better, and most jarring cases are fixed by making pet evade when the owner evades.

Still unsure what the best approach AssistDelayEvent::Execute is to initiate combat, but so far it seems to work fine like this so ...

T1ti commented on src/server/game/Entities/Creature/Creature.cpp:
```diff
@@ -3743,3 +3743,19 @@ void Creature::UpdateLeashExtensionTime()
+
+    if (clevel >= 1 && clevel <= 29)
+        return 11;
+    else if (clevel >= 30 && clevel <= 39)
```
You don't need to recheck min level everytime, if something is below 30 first statement will catch it.
Also I'd recommend using a switch, it's faster and cleaner

zana244 replied:
Done, although I kept the if-else since I've no idea how to do a clean switch for a range of values 😆. Should look a bit cleaner. Ty for the advice
